lib/mobile_app_backend/global_data_cache.ex

```diff
@@ -24,40 +26,132 @@ defmodule MobileAppBackend.GlobalDataCache do
     defstruct [:key, :update_ms]
   end
 
-  def default_key, do: __MODULE__
+  @callback default_key :: term()
+  @callback get_data(key()) :: data()
 
-  def start_link(opts) do
+  @doc """
+  Get the id of all routes that serve the given stops
+  """
+  @callback route_ids_for_stops([Stop.id()], key()) :: :error | [Route.id()]
+
+  def start_link(opts \\ []) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).start_link(opts)
+  end
+
+  def init(opts \\ []) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).init(opts)
+  end
+
+  def default_key do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).default_key()
+  end
+
+  def get_data(key \\ default_key()) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).get_data(key)
+  end
+
+  def route_ids_for_stops(stop_ids, key \\ default_key()) do
+    Application.get_env(
+      :mobile_app_backend,
+      MobileAppBackend.GlobalDataCache.Module,
+      MobileAppBackend.GlobalDataCache.Impl
+    ).route_ids_for_stops(stop_ids, key)
+  end
+end
+
+defmodule MobileAppBackend.GlobalDataCache.Impl do
+  use GenServer
+  alias MBTAV3API.{JsonApi, Repository}
+  alias MobileAppBackend.GlobalDataCache
+  alias MobileAppBackend.GlobalDataCache.State
+
+  @behaviour GlobalDataCache
+
+  @impl true
+  def default_key, do: MobileAppBackend.GlobalDataCache
+
+  @spec start_link(keyword()) :: :ignore | {:error, any()} | {:ok, pid()}
+  def start_link(opts \\ []) do
     opts = Keyword.merge([key: default_key()], opts)
-    GenServer.start_link(__MODULE__, opts)
+    GenServer.start_link(GlobalDataCache, opts)
   end
 
-  @spec get_data(key()) :: data()
+  @impl true
   def get_data(key \\ default_key()) do
     :persistent_term.get(key, nil) || update_data(key)
```

Comment:
Maybe this `update_data` should be moved into a `GenServer.call` or something so that if there are a dozen simultaneous calls to `get_data/1` before data is loaded they don't each call `update_data/1`.

Reply:
I'd be somewhat worried about putting it into `GenServer.call` since if it is slow for the first user, subsequent user requests will all fail too. 

I think the best bet is making sure that the data is populated first & removing the call to `update_data` from `get_data`. 

Doing that asynchronously via the scheduled checks & preventing user traffic via a healthcheck seems like the safest approach for that to me - if the global data can't load immediately for some reason, it seems cleaner to continue trying to re-fetch without crashing. Maybe I'm overly wary of crashes though. In any case, I think resolving that mechanism could be part of a separate PR so that the polling is in place

Reply:
@boringcactus I'm going to break this PR up into separate ones so that we can clear out the immediate problem of setting up the timed refresh of this data. 

I'm in favor of the health check approach over fetching the data in `init` in failing. In speaking to Paul about it (since he might be pitching in for that change anyway), it has the advantage over the `init` of faster deploys in the case that the request fails. Skate and API take that approach as well.

Reply:
I don't think Glides has any particular application-logic-specific health checks in its `/_health`, which is probably why I hadn't thought of it, but I guess if that's common we may as well do it here. Good catch on fixing the refresh first - I had completely lost track of that in the context of the TestFlight public beta.